adapters/handlers/graphql/local/common_filters/near_vector.go

```diff
@@ -65,39 +64,69 @@ func ExtractNearVector(source map[string]interface{}, targetVectorsFromOtherLeve
-			args.VectorPerTarget = map[string][]float32{"": vector}
+			args.Vectors = [][]float32{vector}
 		} else {
-			args.VectorPerTarget = make(map[string][]float32, len(targetVectors))
-			for _, target := range targetVectors {
-				args.VectorPerTarget[target] = vector
+			args.Vectors = make([][]float32, len(targetVectors))
+			for i := range targetVectors {
+				args.Vectors[i] = vector
 			}
 		}
 	}
 
 	if okVecPerTarget {
-		targets := make([]string, 0, len(vectorPerTarget))
-		for target := range vectorPerTarget {
-			targets = append(targets, target)
-		}
-
+		var vectors [][]float32
+		// needs to handle the case of targetVectors being empty (if you only provide a near vector with targets)
 		if len(targetVectors) == 0 {
+			targets := make([]string, 0, len(vectorPerTarget))
+			vectors = make([][]float32, 0, len(vectorPerTarget))
+
+			for target := range vectorPerTarget {
+				single, ok := vectorPerTarget[target].([]float32)
+				if ok {
+					vectors = append(vectors, single)
+					targets = append(targets, target)
+				} else {
+					multiple, okMulti := vectorPerTarget[target].([][]float32)
+					if !okMulti {
+						return searchparams.NearVector{}, nil,
+							fmt.Errorf("vectorPerTarget should be a map with strings as keys and list of floats or list of lists of floats as values. Received %T", vectorPerTarget[target])
+					}
+					for j := range multiple {
+						vectors = append(vectors, multiple[j])
+						targets = append(targets, target)
+					}
+				}
+			}
 			args.TargetVectors = targets
 		} else {
-			if len(targetVectors) != len(targets) {
-				return searchparams.NearVector{}, nil,
-					fmt.Errorf("number of targets (%d) does not match number of vectors (%d)", len(targetVectors), len(targets))
-			}
-			sort.Strings(targetVectors)
-			sort.Strings(targets)
+			vectors = make([][]float32, len(targetVectors))
+			handled := make(map[string]struct{})
 			for i, target := range targetVectors {
-				if target != targets[i] {
-					return searchparams.NearVector{}, nil,
-						fmt.Errorf("targets do not match: %s != %s", target, targets[i])
+				if _, ok := handled[target]; ok {
+					continue
+				} else {
+					handled[target] = struct{}{}
 				}
-			}
+				vectorPerTargetParsed, ok := vectorPerTarget[target]
+				if !ok {
+					return searchparams.NearVector{}, nil, fmt.Errorf("vectorPerTarget for target %s is not provided", target)
+				}
+				if vectorIn, ok := vectorPerTargetParsed.([]float32); ok {
+					vectors[i] = vectorIn
+				} else if vectorsIn, ok := vectorPerTargetParsed.([][]float32); ok {
+					for j, w := range vectorsIn {
 
+						if i+j >= len(targetVectors) || targetVectors[i+j] != target {
+							return searchparams.NearVector{}, nil, fmt.Errorf("target %s is not in the correct order", target)
+						}
+						vectors[i+j] = w
+					}
+				} else {
+					return searchparams.NearVector{}, nil, fmt.Errorf("weight for target %s is not a float, got %v", target, vectorPerTargetParsed)
+				}
```

Comment:
I find this code hard to understand reading it as is, especially the statement on L119. Would it be possible to add some comments around this aspect or as part of a docstring describing its action?

Reply:
is the reworked code better to understand?